enhancements/rhcos/split-rhcos-into-layers.md

```diff
@@ -0,0 +1,428 @@
+
+* As an OKD contributor, I want to be able to use CentOS Stream-based boot images, so that I don't have to keep pivoting from Fedora CoreOS boot images into CentOS Stream everytime.
+
+* As an OKD contributor, I want to be able to share node image building processes with OCP, so that I can reduce the maintenance load.
+
+* (Longer-term) As a product manager, I want image mode for RHEL certification and support guarantees to automatically apply to OpenShift clusters.
+
+### Goals
+
+- Build RHCOS/SCOS on top of rhel-bootc/centos-bootc
+- Share efforts, especially CI and docs, with the larger rhel/centos-bootc project
+- Maintain a single RHCOS stream per RHEL stream
+- Build the OpenShift node image as a layered image on top of RHCOS/SCOS
+- Build the OpenShift node image using Prow (in CI) and Konflux (in prod)
+
+### Non-Goals
+
+- Change the cluster installation flow. It should remain the same whether IPI/UPI/AI/etc...
+- Introduce cluster administrator-visible changes. This change should be transparent to
+  admistrators. CoreOS layering instructions should keep working as is, but documentation
+  should ideally be reworked to leverage rhel-bootc docs more.
+- Remove Ignition. While image mode for RHEL lessens the need for Ignition, it remains a
+  crucial part of the provisioning flow and will stick around for the foreseeable future.
+
+## Proposal
+
+The overall proposal consists of two parts. In the first part, we will focus on moving
+the OCP-specific packages out of RHCOS and into its own layer (the node layer). In the second
+part, we'll change RHCOS itself to become a layered image (the CoreOS layer) built on top
+of the rhel-bootc work (the bootc layer).
+
+As mentioned in the summary, this enhancement is focused on detailing the first part only.
+Firstly, because it's the one we'll start with, and secondly, because it's the one that
+will most impact OpenShift. The second part will require changes that are more contained
+to the CoreOS and ART teams. That said, we felt it important to introduce them together
+in this enhancement to provide a larger picture.
+
+Diving into the first part then, at a high-level this is split into three phases.
+
+#### Phase 1: Using RHEL/CentOS Stream-only bootimages
+
+In this phase, the goal is to have the OpenShift installer use bootimages containing only
+RHEL/CentOS Stream content.
+
+To do this, we will start building two new streams in the RHCOS pipeline containing only pure
+RHEL/CentOS Stream content (let's call these the "pure RHEL stream" and "pure CentOS stream").
+Those streams will also be building the usual bootimages and uploading them to cloud.
+
+Once we have these bootimages, we can better start adapting components that will need it
+to account for the lack of OpenShift components in the bootimages. Likely suspects here are
+any components involved in the bringup of the cluster (installer, Assisted Installer, MCO, etc...).
```

Comment:
We have three choices broadly:

- Per above, pull and reboot into OCP container image, what is already done in installer today for OCP.  Note a subtle but very important point today...OCP sometimes ships kernels outside of RHEL cycles, and while that's not *usually* relevant for bootstrap, I wouldn't say it never would be
- Special case bootstrap to just e.g. `bootc usroverlay + dnf install`
- Run the installation process in a container (IMO...this is actually best)

Reply:
That reads to me like a next step after this enhancement? Ideally we don't change the cluster installation flow (yet) and complete this transition first before looking at other installation options?

Reply:
> That reads to me like a next step after this enhancement? Ideally we don't change the cluster installation flow (yet) and complete this transition first before looking at other installation options?

If the bootimages are changed to not contain OCP content as the enhancement says, then things will definitely need to change in the install process.

Or, that bit could be moved to a later phase, which also seems viable to me (but I think also lowers the value of things, because being able to reuse the same bootimages for *multiple* OCP versions would be quite a big improvement).

Reply:
> Note a subtle but very important point today...OCP sometimes ships kernels outside of RHEL cycles, and while that's not usually relevant for bootstrap, I wouldn't say it never would be

Can you expand on this? E.g. FIPS/crypto-related things? I could imagine detecting those cases where we should reboot and only do so then. Though in the common case it'd be very desirable to try to avoid rebooting.

> Run the installation process in a container (IMO...this is actually best)

Doesn't that imply trying to run the kubelet containerized? Not sure how supported that is nowadays.

> Special case bootstrap to just e.g. `bootc usroverlay + dnf install`

Yeah, definitely a potential path forward that's not too complex. We could ship the kubelet RPM in the extensions container image, which is already in the payload.

Hmm, or another approach is to just copy the kubelet out of `rhel-coreos`, since it's almost static anyway. There's still glibc, but I don't think mismatch issues would be a concern here (this isn't the "new worker node" flow with arbitrarily old bootimages, here we're doing an install so we should be guaranteed to be using matching images + it's a known combination that can be extensively tested in CI).

Reply:
> Doesn't that imply trying to run the kubelet containerized? Not sure how supported that is nowadays.

Yes, but the important thing here is that we don't need any non-core kubelet functionality.  In particular, CSI (storage) drivers running containerized I think were the #1 problem.  But we don't need any of that, it just needs to run pods.

There is also the angle that nowadays we could just pass pod definitions directly to podman...which I don't believe existed or was really fleshed out when OCP 4.0 was being designed.  That may even be simplest.

Reply:
I added this as an open question, but keeping this thread unresolved since that's where all the context.

Reply:
> There is also the angle that nowadays we could just pass pod definitions directly to podman...which I don't believe existed or was really fleshed out when OCP 4.0 was being designed. That may even be simplest.

`podman play kube` definitely sounds like the simplest way to get the current bootstrap flow working without kubelet.

Reply:
> Hmm, or another approach is to just copy the kubelet out of rhel-coreos, since it's almost static anyway. There's still glibc, but I don't think mismatch issues would be a concern here

This might lead bugs such as: e.g. https://issues.redhat.com/browse/MGMT-16705 ( in this case the kubelet is incompatible because it's doing late-binding hence the node was booted with a default RHCOS and the not RHCOS of the release, yet the kubelet.conf is the one for the release).

Reply:
> Likely suspects here are any components involved in the bringup of the cluster

I think it would be more helpful to focus on stages rather than components. There are 3 that will need to work. In reverse order:

1. Booting a host as a node after it is installed. Presumably this is a fairly straightforward matter of having it boot the OCP layer from the release payload.
2. Running the bootstrap node. This normally depends on OCP components. In the Assisted/ABI case, it necessarily runs on the base image, and we cannot stop services that were already running on the host prior to bootstrap starting.
3. The Assisted discovery ISO/ABI agent ISO. I believe this is a no-op, as neither of them should depend on OCP components today.

Reply:
> > Hmm, or another approach is to just copy the kubelet out of rhel-coreos, since it's almost static anyway. There's still glibc, but I don't think mismatch issues would be a concern here
> 
> This might lead bugs such as: e.g. [issues.redhat.com/browse/MGMT-16705](https://issues.redhat.com/browse/MGMT-16705) ( in this case the kubelet is incompatible because it's doing late-binding hence the node was booted with a default RHCOS and the not RHCOS of the release, yet the kubelet.conf is the one for the release).

Is this something we should support though? Having the bootstrap process also have to worry about different bootimage versions instead of only the blessed, heavily CI-tested, version in `rhcos.json` doesn't sound great.

Reply:
> Running the bootstrap node. This normally depends on OCP components. In the Assisted/ABI case, it necessarily runs on the base image, and we cannot stop services that were already running on the host prior to bootstrap starting.

What does ABI refer to here?

Can you expand/link to more information re. the "we cannot stop services that were already running on the host prior to bootstrap starting"? Don't we own the bootstrap node?

Reply:
ABI = agent-based installer

In both assisted and ABI installs, the bootstrap-in-place runs inside the live ISO that is already running the agent. I believe assisted-service doesn't take well to the agent being restarted, but ABI has even tighter requirements: assisted-service itself is running on the bootstrap node, so interrupting it will break the entire installation process.

Reply:
With https://github.com/openshift/installer/pull/8742 and https://github.com/openshift/assisted-installer/pull/899, I think we have a path forward now on this. Resolving.